File: clients/java/src/test/java/io/volant/GroupsAdminTest.java
package io.volant;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.InetSocketAddress;
import java.net.ServerSocket;
import java.net.Socket;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicReference;
import org.junit.jupiter.api.Test;

/** ListGroups / DescribeGroup client tests against a fake native server. */
class GroupsAdminTest {
    @Test
    void listGroupsEmptyAndStable() throws Exception {
        try (GroupAdminServer srv = new GroupAdminServer(0)) {
            try (Client c = Client.connect("127.0.0.1", srv.port, 5_000)) {
                List<Codec.GroupListing> groups = c.listGroups();
                assertEquals(2, groups.size());
                Map<String, Codec.GroupListing> byId = new HashMap<>();
                for (Codec.GroupListing g : groups) {
                    byId.put(g.groupId, g);
                }
                assertEquals(Codec.GROUP_STATE_EMPTY, byId.get("g2").state);
                assertEquals(0, byId.get("g2").memberCount);
                assertEquals(0, byId.get("g2").generation);
                assertEquals(Codec.GROUP_STATE_STABLE, byId.get("g1").state);
                assertEquals(2, byId.get("g1").memberCount);
                assertEquals(5, byId.get("g1").generation);
            }
            srv.assertOk();
            assertEquals(Collections.singletonList(Codec.OP_LIST_GROUPS), srv.opcodes);
        }
    }

    @Test
    void describeGroupMembersAndAssignment() throws Exception {
        try (GroupAdminServer srv = new GroupAdminServer(0)) {
            try (Client c = Client.connect("127.0.0.1", srv.port, 5_000)) {
                DescribeGroupResult desc = c.describeGroup("cg-1");
                assertEquals("cg-1", desc.groupId);
                assertEquals(3, desc.generation);
                assertEquals(1, desc.members.size());
                Codec.GroupMemberInfo m = desc.members.get(0);
                assertEquals("m-a", m.memberId);
                assertEquals(Collections.singletonList("events"), m.topics);
                assertEquals(2, m.assignment.size());
                assertEquals("events", m.assignment.get(0).topic);
                assertEquals(0, m.assignment.get(0).partition);
                assertEquals(2, m.assignment.get(1).partition);
            }
            srv.assertOk();
            assertEquals("cg-1", srv.described.get());
        }
    }

    @Test
    void describeGroupNotFoundRaises() throws Exception {
        try (GroupAdminServer srv = new GroupAdminServer(2)) {
            try (Client c = Client.connect("127.0.0.1", srv.port, 5_000)) {
                BrokerException ex = assertThrows(BrokerException.class, () -> c.describeGroup("missing"));
                assertEquals(2, ex.code);
                assertEquals("describe_group", ex.op);
            }
            srv.assertOk();
            assertEquals("missing", srv.described.get());
        }
    }

    private static final class GroupAdminServer implements AutoCloseable {
        final int port;
        final List<Integer> opcodes = Collections.synchronizedList(new java.util.ArrayList<>());
        final AtomicReference<String> described = new AtomicReference<>();
        private final int describeError;
        private final ServerSocket listen;
        private final Thread thread;
        private final AtomicReference<Exception> error = new AtomicReference<>();
        private final CountDownLatch done = new CountDownLatch(1);

        GroupAdminServer(int describeError) throws IOException {
            this.describeError = describeError;
            ServerSocket ss = new ServerSocket();
            ss.setReuseAddress(true);
            ss.bind(new InetSocketAddress("127.0.0.1", 0));
            ss.setSoTimeout(5_000);
            this.listen = ss;
            this.port = ss.getLocalPort();
            this.thread = new Thread(
                    () -> {
                        try (Socket conn = listen.accept()) {
                            handle(conn);
                        } catch (Exception e) {
                            error.set(e);
                        } finally {
                            done.countDown();
                        }
                    },
                    "volant-groups");
            this.thread.setDaemon(true);
            this.thread.start();
        }

        void assertOk() throws Exception {
            if (!done.await(5, TimeUnit.SECONDS)) {
                throw new IOException("server did not finish");
            }
            Exception e = error.get();
            if (e != null) {
                throw e;
            }
        }

        @Override
        public void close() {
            try {
                listen.close();
            } catch (IOException ignored) {
                // best-effort
            }
            try {
                thread.join(2_000);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        }

        private void handle(Socket conn) throws Exception {
            conn.setSoTimeout(5_000);
            InputStream in = conn.getInputStream();
            OutputStream out = conn.getOutputStream();
            byte[] buf = new byte[0];
            while (true) {
                Frame.Decode d = Frame.tryDecode(buf);
                if (d.frame == null) {
                    byte[] tmp = new byte[4096];
                    int n = in.read(tmp);
                    if (n < 0) {
                        return;
                    }
                    byte[] next = new byte[buf.length + n];
                    System.arraycopy(buf, 0, next, 0, buf.length);
                    System.arraycopy(tmp, 0, next, buf.length, n);
                    buf = next;
                    continue;
                }
                buf = d.rest;
                opcodes.add(d.frame.opcode);
                if (d.frame.opcode == Codec.OP_LIST_GROUPS) {
                    byte[] payload = Codec.encodeListGroupsResponse(
                            new Codec.ListGroupsResponse(
                                    0,
                                    Arrays.asList(
                                            new Codec.GroupListing("g2", Codec.GROUP_STATE_EMPTY, 0, 0),
                                            new Codec.GroupListing("g1", Codec.GROUP_STATE_STABLE, 2, 5))));
                    out.write(Frame.encode(Codec.OP_LIST_GROUPS_RESPONSE, d.frame.correlationId, payload));
                    out.flush();
                    continue;
                }
                if (d.frame.opcode == Codec.OP_DESCRIBE_GROUP) {
                    String groupId = Codec.decodeDescribeGroupRequest(d.frame.payload).groupId;
                    described.set(groupId);
                    Codec.DescribeGroupResponse resp;
                    if (describeError != 0) {
                        resp = new Codec.DescribeGroupResponse(describeError, groupId, 0, Collections.emptyList());
                    } else {
                        resp = new Codec.DescribeGroupResponse(
                                0,
                                groupId,
                                3,
                                Collections.singletonList(
                                        new Codec.GroupMemberInfo(
                                                "m-a",
                                                Collections.singletonList("events"),
                                                Arrays.asList(
                                                        new Codec.Assignment("events", 0),
                                                        new Codec.Assignment("events", 2)))));
                    }
                    byte[] payload = Codec.encodeDescribeGroupResponse(resp);
                    out.write(Frame.encode(Codec.OP_DESCRIBE_GROUP_RESPONSE, d.frame.correlationId, payload));
                    out.flush();
                    if (describeError != 0) {
                        return;
                    }
                    continue;
                }
                return;
            }
        }
    }
}

File: clients/go/groups_admin_test.go
package volant_test

import (
	"errors"
	"net"
	"testing"
	"time"

	volant "github.com/volant-mq/volant/clients/go"
	"github.com/volant-mq/volant/clients/go/codec"
	"github.com/volant-mq/volant/clients/go/frame"
)

type groupAdminResult struct {
	opcodes   []uint16
	described string
	err       error
}

func serveGroupAdmin(t *testing.T, describeError uint16) (addr string, got *groupAdminResult, stop func()) {
	t.Helper()
	ln, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		t.Fatal(err)
	}
	res := &groupAdminResult{}
	done := make(chan struct{})
	go func() {
		defer close(done)
		conn, err := ln.Accept()
		if err != nil {
			res.err = err
			return
		}
		defer conn.Close()
		_ = conn.SetDeadline(time.Now().Add(5 * time.Second))
		buf := make([]byte, 0, 4096)
		tmp := make([]byte, 4096)
		for {
			f, rest, err := frame.TryDecode(buf)
			if err != nil {
				res.err = err
				return
			}
			if f == nil {
				n, err := conn.Read(tmp)
				if n > 0 {
					buf = append(buf, tmp[:n]...)
				}
				if err != nil {
					res.err = err
					return
				}
				continue
			}
			buf = append([]byte(nil), rest...)
			res.opcodes = append(res.opcodes, f.Opcode)
			switch f.Opcode {
			case codec.OpListGroups:
				payload, err := codec.EncodeListGroupsResponse(codec.ListGroupsResponse{
					ErrorCode: 0,
					Groups: []codec.GroupListing{
						{GroupID: "g2", State: codec.GroupStateEmpty, MemberCount: 0, Generation: 0},
						{GroupID: "g1", State: codec.GroupStateStable, MemberCount: 2, Generation: 5},
					},
				})
				if err != nil {
					res.err = err
					return
				}
				raw, err := frame.Encode(codec.OpListGroupsResponse, f.CorrelationID, payload)
				if err != nil {
					res.err = err
					return
				}
				if _, err := conn.Write(raw); err != nil {
					res.err = err
					return
				}
			case codec.OpDescribeGroup:
				req, err := codec.DecodeDescribeGroupRequest(f.Payload)
				if err != nil {
					res.err = err
					return
				}
				res.described = req.GroupID
				var resp codec.DescribeGroupResponse
				if describeError != 0 {
					resp = codec.DescribeGroupResponse{
						ErrorCode: describeError,
						GroupID:   req.GroupID,
					}
				} else {
					resp = codec.DescribeGroupResponse{
						ErrorCode:  0,
						GroupID:    req.GroupID,
						Generation: 3,
						Members: []codec.GroupMemberInfo{
							{
								MemberID: "m-a",
								Topics:   []string{"events"},
								Assignment: []codec.Assignment{
									{Topic: "events", Partition: 0},
									{Topic: "events", Partition: 2},
								},
							},
						},
					}
				}
				payload, err := codec.EncodeDescribeGroupResponse(resp)
				if err != nil {
					res.err = err
					return
				}
				raw, err := frame.Encode(codec.OpDescribeGroupResponse, f.CorrelationID, payload)
				if err != nil {
					res.err = err
					return
				}
				if _, err := conn.Write(raw); err != nil {
					res.err = err
					return
				}
				if describeError != 0 {
					return
				}
			default:
				return
			}
		}
	}()
	return ln.Addr().String(), res, func() {
		_ = ln.Close()
		select {
		case <-done:
		case <-time.After(2 * time.Second):
		}
	}
}

func TestListGroupsEmptyAndStable(t *testing.T) {
	addr, got, stop := serveGroupAdmin(t, 0)
	defer stop()
	c, err := volant.DialTimeout(addr, 5*time.Second)
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()
	groups, err := c.ListGroups()
	if err != nil {
		t.Fatal(err)
	}
	if len(groups) != 2 {
		t.Fatalf("groups %#v", groups)
	}
	byID := map[string]volant.GroupListing{}
	for _, g := range groups {
		byID[g.GroupID] = g
	}
	if byID["g2"].State != volant.GroupStateEmpty || byID["g2"].MemberCount != 0 || byID["g2"].Generation != 0 {
		t.Fatalf("empty %#v", byID["g2"])
	}
	if byID["g1"].State != volant.GroupStateStable || byID["g1"].MemberCount != 2 || byID["g1"].Generation != 5 {
		t.Fatalf("stable %#v", byID["g1"])
	}
	if len(got.opcodes) != 1 || got.opcodes[0] != codec.OpListGroups {
		t.Fatalf("opcodes %#v", got.opcodes)
	}
	if got.err != nil {
		t.Fatal(got.err)
	}
}

func TestDescribeGroupMembersAndAssignment(t *testing.T) {
	addr, got, stop := serveGroupAdmin(t, 0)
	defer stop()
	c, err := volant.DialTimeout(addr, 5*time.Second)
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()
	desc, err := c.DescribeGroup("cg-1")
	if err != nil {
		t.Fatal(err)
	}
	if desc.GroupID != "cg-1" || desc.Generation != 3 || len(desc.Members) != 1 {
		t.Fatalf("desc %#v", desc)
	}
	m := desc.Members[0]
	if m.MemberID != "m-a" || len(m.Topics) != 1 || m.Topics[0] != "events" {
		t.Fatalf("member %#v", m)
	}
	if len(m.Assignment) != 2 || m.Assignment[0].Topic != "events" || m.Assignment[1].Partition != 2 {
		t.Fatalf("assignment %#v", m.Assignment)
	}
	if got.described != "cg-1" {
		t.Fatalf("described %q", got.described)
	}
	if got.err != nil {
		t.Fatal(got.err)
	}
}

func TestDescribeGroupNotFoundRaises(t *testing.T) {
	addr, got, stop := serveGroupAdmin(t, 2)
	defer stop()
	c, err := volant.DialTimeout(addr, 5*time.Second)
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()
	_, err = c.DescribeGroup("missing")
	if err == nil {
		t.Fatal("expected not found")
	}
	var be *codec.BrokerError
	if !errors.As(err, &be) || be.Code != 2 || be.Op != "describe_group" {
		t.Fatalf("err=%v", err)
	}
	if got.described != "missing" {
		t.Fatalf("described %q", got.described)
	}
}
